flang/lib/Optimizer/Builder/IntrinsicCall.cpp

```diff
@@ -2486,12 +2487,34 @@ IntrinsicLibrary::genAssociated(mlir::Type resultType,
     // In both cases, ASSOCIATED should be false if POINTER is NULL.
     return builder.create<mlir::arith::AndIOp>(loc, sameTarget, notNull);
   }
+
+  const fir::ExtendedValue &target = args[1];
+
+  if (!args[0].getBoxOf<fir::MutableBoxValue>()) {
```

Comment:
I would not expect any change here. IntrinsicCall.cpp should not depend on HLFIR.
Usually, these kind of issues are packaging issues related to the mlir::Value to fir::ExtendedValue implicit ctor.

Here, I suspect you need to change `getArgument` https://github.com/llvm/llvm-project/blob/a72a90677d2b320e3bca553698e99143034387d9/flang/lib/Lower/ConvertCall.cpp#L1793 to `return Fortran::lower::translateToExtendedValue(loc, builder, actual, callContext.stmtCtx);` to get the correct packaging. My bad, I wrote it with simple scalars in mind...

Then, you should not need any changes in both IntrinsicCall.cpp and CustomIntrinsicCall.cpp.

Reply:
Ah I see now! Yeah I should have figured that out when I saw the weird emboxing. I'll update that as well and it will look much less hacky.

Reply:
It indeed look much better with your advice! Thanks!